text/0000-futures.md

```diff
@@ -0,0 +1,813 @@
+for helping detect bugs:
+
+```rust
+/// Marks the current thread as being within the dynamic extent of an
+/// executor.
+///
+/// Executor implementations should call this function before beginning to
+/// execute a tasks, and drop the returned `Enter` value after completing
+/// task execution:
+///
+/// ```rust
+/// let enter = enter().expect("...");
+/// /* run task */
+/// drop(enter);
+/// ```
+///
+/// Doing so ensures that executors aren't accidentally invoked in a nested fashion.
+///
+/// # Error
+///
+/// Returns an error if the current thread is already marked, in which case the
+/// caller should panic with a tailored error message.
+pub fn enter() -> Result<Enter, EnterError>
+```
+
+As stated in the doc comment, the expectation is that all executors will wrap
+their task execution within an `enter` to detect inadvertent nesting.
+
+### Task contexts
+
+All tasks are executed with two pieces of contextual information:
+
+- A `Waker` for waking the task up later on.
+- An executor, which is the "default" place to spawn further tasks.
+
+Notably, this list does *not* include task-local data; that can be addressed
+externally, as we'll see in a later section.
+
+The `core::task::Context` type gathers the (stack-rooted) contextual information
+together, and is passed by mutable reference to all polling functions:
+
+```rust
+/// Information about the currently-running task.
+///
+/// Contexts are always tied to the stack, since they are set up specifically
+/// when performing a single `poll` step on a task.
+pub struct Context<'a> { .. }
+
+impl<'a> Context<'a> {
+    /// Note: this signature is future-proofed for `E: Executor` later.
+    pub fn new<E>(waker: &'a Waker, executor: &'a mut E) -> Context<'a>
+        where E: BoxExecutor;
+
+    /// Get the `Waker` associated with the current task.
+    pub fn waker(&self) -> &Waker
+
+    /// Spawn a future onto the default executor.
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if the default executor is unable to spawn.
+    /// To handle executor errors, use the `executor` method instead.
+    pub fn spawn(&mut self, f: impl Future<Item = ()> + 'static + Send);
+
+    /// Get the default executor associated with this task.
+    ///
+    /// This method is useful primarily if you want to explicitly handle
+    /// spawn failures.
+    ///
+    /// NB: this will remain unstable until the final `Executor` trait is ready.
+    pub fn executor(&mut self) -> &mut BoxExecutor;
+}
+```
+
+Note that the `spawn` method here will box until `Executor` is added.
+
+## `core::future` module
+
+With all of the above task infrastructure in place, defining `Future` is
+straightforward.  Like `Iterator`, the `Future` trait has a single required
+method, `poll`, and a large number of provided methods (called "adapters" or
+"combinators"). We'll look first at `poll`:
+
+```rust
+pub trait Future {
+    /// The type of value produced on completion.
+    type Item;
```

Comment:
We were discussing this last week: I believe `Output` is more in line here with existing conventions (e.g. `Fn` traits, operator overloading traits), and `Item` would be more appropriate for `Stream` (by association with `Iterator`).

Reply:
`Outcome`. Just kidding

Reply:
@withoutboats We could do some sort of `Future(T) -> R` deal for all traits of the form:
```rust
trait TraitName<Arg> {
    type Output;
    // other stuff does not matter...
}
```

This might have a knock on benefit for people who define their own "Fn-like" traits.

Personally I think 2 more characters for a more apt name is the right way to go in this instance.

Reply:
I like the `Fn => Future`, `Iterator => Stream` correspondence, and I think using `Output` will avoid confusion in places where futures/streams are being mixed.

Reply:
@withoutboats I am in strong support of keeping it `Item` for the reason you mentioned. As someone who works w/ futures all day, I type `Item` constantly. The shorter name makes a big difference here.

The fn analogy doesn't make sense. a future is **not** a function. It is a value that will complete in the future.

`Item` matches `Iterator` which is closer to what a future is.

Reply:
> The fn analogy doesn't make sense. a future is not a function. It is a value that will complete in the future.

The analogy does make sense: both a `Fn` and a `Future` produce a single *`Output`*. Both an `Iterator` and a `Stream` produce multiple *`Items`*

Reply:
To make my position a bit clearer: I think `Item` made more sense when we had something like `Result<Self::Item, Self::Error>`, since it was "the item in the result".

Reply:
@withoutboats 
> For Fn traits this doesn't matter, they have -> syntax, but Future, like Iterator, will often be bound Future<Item = ?>.

We could make `->` sugar work for the `Future` trait, I suppose:

```rust
fn foo() -> impl Future -> u32 {
    println!("Foo called");
    async {
        println!("Foo polled");
        5
    }
}

fn serve(f: impl FnMut(Request) -> impl Future -> Response) {
    // serve http requests by calling `f`
    ....
}
```